generator/internal/http_option_utils.cc

```diff
@@ -242,21 +252,37 @@ ParseHttpExtension(google::protobuf::MethodDescriptor const& method) {
                      << ": google::api::HttpRule not handled";
   }
 
-  static std::regex const kUrlPatternRegex(R"((.*)\{(.*)=(.*)\}(.*))");
-  std::smatch match;
-  if (!std::regex_match(url_pattern, match, kUrlPatternRegex)) {
+  static std::regex const kImplicitNamedFieldUrlPatternRegex(
+      R"((.*)\{(.*)\}(.*))");
+  std::smatch implicit_match;
+
+  // The implicit named field regex matches both the implicit and explicit
+  // syntax. So we can use the implicit match to determine if either are
+  // present.
+  if (!std::regex_match(url_pattern, implicit_match,
+                        kImplicitNamedFieldUrlPatternRegex)) {
     return HttpSimpleInfo{info.http_verb, url_pattern, http_rule.body()};
   }
-  info.url_path = match[0];
-  info.request_field_name = match[2];
-  info.url_substitution = match[3];
-  info.body = http_rule.body();
-  info.path_prefix = match[1];
-  info.path_suffix = match[4];
+
+  static std::regex const kExplicitNamedFieldUrlPatternRegex(
+      R"((.*)\{(.*)=(.*)\}(.*))");
+  std::smatch explicit_match;
+
+  if (std::regex_match(url_pattern, explicit_match,
+                       kExplicitNamedFieldUrlPatternRegex)) {
+    info.url_path = explicit_match[0];
+    info.body = http_rule.body();
+    info.path_suffix = explicit_match[4];
+  } else {
+    info.url_path = implicit_match[0];
+    info.body = http_rule.body();
+    info.path_suffix = implicit_match[3];
+  }
 
   std::size_t current = 0;
-  for (auto open = url_pattern.find('{'); open != std::string::npos;
-       open = url_pattern.find('{', current)) {
+  auto open = url_pattern.find('{');
+  info.path_prefix = url_pattern.substr(current, open - current);
+  for (; open != std::string::npos; open = url_pattern.find('{', current)) {
```

Comment:
I think there is a code smell here.  We are mixing regexp and explicit parsing.  That suggests we are doing something odd, like the grammar we want to parse is too complex for regex, but we don't want to write a full parser.

🤷